### DISCUSSION TITLE
Apollo Server and Envelop

### DISCUSSION THREAD

**Discussion by borekb** [apollographql/apollo-server#5541] [Ideas] [answered]
Starting this discussion based on https://twitter.com/UriGoldshtein/status/1418529860578906112 by @Urigo:

> We shared Envelop with Apollo a 2 months ago. 
They loved the idea but said they are focusing on releasing v3 and open for this change in the future.
Would be cool if someone started a thread on their repo and we could start discussing how this could look like now that v3 is out

I'm indeed interested in know if it's realistic for Apollo Server to "rebuild itself on top of Envelop" (at least the way I understand it is that Envelop is not _added_ to a project; the project needs to be built on top of it). Generally, it would be nice to be able to add Envelop plugins to Apollo Server, especially if the ecosystem grows in the future.

**Comment by glasser** [comment, marked as answer]
As Uri said, Envelop is pretty cool!

I think right now the likelihood of fully rewriting Apollo Server on top of it is low. We finally with AS3 got rid of our old "extensions" plugin system so that we're back to only having one plugin system again, and having two separate ones in AS3 might be a little much.  But there might be ways that we can make the systems more interchangeable with each other.

**Reply by borekb** [reply]
Thanks for the quick reply, this is perfectly understandable. Congrats to the Apollo Server 3.0 release BTW!

**Reply by dotansimha** [reply]
Hi @glasser ! Congrats on v3 of `apollo-server` 🎉 

First, I want to note that currently, Apollo-Server and `envelop` are working together, you can check the example here: https://github.com/dotansimha/envelop/tree/main/examples/apollo-server#envelop-example-with-apollo-server 

But the current setup prevents you from using [all `envelop` plugins](https://www.envelop.dev/plugins), and you can only use some of them. 

But the good news is that I think the effort of adding full support for all of `envelop`'s plugins to the existing implementation of `apollo-server` might even be simpler than it looks, and definitely doesn't need to include a "full rewrite". 

Since `envelop` only cares about wrapping the GraphQL pipeline functions, it should be possible to add it today, just by extending the existing API of `ApolloServer`.

Here are the steps that I had in mind:

1. `parse` - today, AS3 doesn't allow to provide a custom `parse` function. I don't see any reason to prevent developers from doing that, especially since AS doesn't really change anything within the `parse` function and calls `graphql#parse` directly (https://github.com/apollographql/apollo-server/blob/12f46fc8e07204d15e4a24a06df594f80c5caed4/packages/apollo-server-core/src/requestPipeline.ts#L234) (so instead of allowing only `parseOptions`, in can also allow passing `parseFn?: typeof parse`)
2. `validate` - same as `parse` (see https://github.com/apollographql/apollo-server/blob/12f46fc8e07204d15e4a24a06df594f80c5caed4/packages/apollo-server-core/src/requestPipeline.ts#L246) 
3. `execute` - Instead of dealing with the custom executor, you can open the API and allow developers to provide a custom `execute` and `subscribe`. Today AS is using the original function from `graphql` (and this can solve other integrations as well, see https://github.com/apollographql/apollo-server/pull/4167 ) 
4. `subscribe` - same as `execute`.

I think allowing to provide these pipeline functions dynamically (based on `req`), or maybe even allowing to provide a custom `GraphQLRequestPipelineConfig` object, will make it super simple to fully support envelop in AS, without changing the entire codebase or rewriting it from scratch (and without introducing any breaking changes to Apollo-Server users).

What do you think?

**Reply by dotansimha** [reply]
Hi @glasser ! 
I saw you recently started working on v4 of `apollo-server`, this is really exciting! Is the plan still to make `apollo-server` more flexible and allow integration with `envelop`?  Is there anything we can do to help?

**Reply by glasser** [reply]
Our focus in v4 is simplifying the connection between Apollo Server and web frameworks to be a well-defined API rather than a bunch of built-in integrations.  The internals of execution and observability and hooks aren't really our focus for this version.

**Reply by to-kn** [reply]
hi @glasser any news in this direction?

**Reply by glasser** [reply]
Nope! Apollo Server v4 was released though.

**Reply by benasher44** [reply]
@dotansimha what limitations are there still with AS4? The envelop docs link this discussion, but it's not clear what limitations exist now in the AS4 world.

**Reply by dotansimha** [reply]
As far as I know, Apollo-Server still doesn't allow you to override things like `validation` function and `parse` function.
